File: source/loadInterfile.m
function output = loadInterfile(filename)
%LOADINTERFILE Loads the input file as a MATLAB matrix with the same
%   dimensions and format type as the original.
%
%   Header file must exist for the file load to work and must contain the
%   number format, number of bytes per pixel and the each individual matrix
%   dimensions. Up to five dimensions are supported natively.
%
%   Supported file types are hdr, img, h33 and i33.
%
% Example:
%   loadInterfile(filename)
%
% Input:
%   filename = Name of the image (.img or .i33) or header file (.hdr or
%   .h33)
%
% See also loadMetaImage
tline = 0;
ll = 1;
M = cell(1,1);
if strcmp(filename(end-2:end),'img') || strcmp(filename(end-2:end),'i33')
    filename = [filename(1:end-3) 'hdr'];
end
fid = fopen(filename);
if fid == -1
    filename = [filename(1:end-3) 'h33'];
    fid = fopen(filename);
    if fid == -1
        error(['Could not find either ' filename ' or ' [filename(1:end-3) 'hdr']])
    end
end
while sum(tline ~= -1) >= 1 || isempty(tline ~= -1)
    tline = fgetl(fid);
    M{ll} = {tline};
    ll = ll + 1;
end
fclose(fid);
M = M(1:end-1);
n_dim1 = 1;
n_dim2 = 1;
n_dim3 = 1;
n_dim4 = 1;
n_dim5 = 1;
machinefmt = 'l';

for kk = 1 : length(M)
    if ~cellfun('isempty',strfind(M{kk},'number format'))
        if ~cellfun('isempty',strfind(M{kk},'float')) || ~cellfun('isempty',strfind(M{kk},'short float'))
            type = 'single';
        elseif ~cellfun('isempty',strfind(M{kk},'double')) || ~cellfun('isempty',strfind(M{kk},'long float'))
            type = 'double';
        elseif ~cellfun('isempty',strfind(M{kk},'unsigned integer'))
            for hh = 1 : length(M)
                if ~cellfun('isempty',strfind(M{hh},'number of bytes per pixel'))
                    apu = cell2mat(M{hh});
                    if strcmp(apu(end), '1')
                        type = 'uint8';
                    elseif strcmp(apu(end), '2')
                        type = 'uint16';
                    elseif strcmp(apu(end), '4')
                        type = 'uint32';
                    elseif strcmp(apu(end), '8')
                        type = 'uint64';
                    else
                        error('Cannot read number of bytes per pixel')
                    end
                    break
                end
            end
        elseif ~cellfun('isempty',strfind(M{kk},'signed integer'))
            for hh = 1 : length(M)
                if ~cellfun('isempty',strfind(M{hh},'number of bytes per pixel'))
                    apu = cell2mat(M{hh});
                    if strcmp(apu(end), '1')
                        type = 'int8';
                    elseif strcmp(apu(end), '2')
                        type = 'int16';
                    elseif strcmp(apu(end), '4')
                        type = 'int32';
                    elseif strcmp(apu(end), '8')
                        type = 'int64';
                    else
                        error('Cannot read number of bytes per pixel')
                    end
                    break
                end
            end
        end
%     elseif ~cellfun('isempty',strfind(M{kk},'number of dimensions'))
%         apu = cell2mat(M{kk});
%         n_dim = str2double(apu(end));
    elseif ~cellfun('isempty',strfind(M{kk},'matrix size[1]'))
        koko = cell2mat(strfind(M{kk},':='));
        apu = cell2mat(M{kk});
        n_dim1 = str2double(apu(koko+2:end));
    elseif ~cellfun('isempty',strfind(M{kk},'matrix size[2]'))
        koko = cell2mat(strfind(M{kk},':='));
        apu = cell2mat(M{kk});
        n_dim2 = str2double(apu(koko+2:end));
    elseif ~cellfun('isempty',strfind(M{kk},'matrix size[3]'))
        koko = cell2mat(strfind(M{kk},':='));
        apu = cell2mat(M{kk});
        n_dim3 = str2double(apu(koko+2:end));
    elseif ~cellfun('isempty',strfind(M{kk},'matrix size[4]'))
        koko = cell2mat(strfind(M{kk},':='));
        apu = cell2mat(M{kk});
        n_dim4 = str2double(apu(koko+2:end));
    elseif ~cellfun('isempty',strfind(M{kk},'matrix size[5]'))
        koko = cell2mat(strfind(M{kk},':='));
        apu = cell2mat(M{kk});
        n_dim5 = str2double(apu(koko+2:end));
    elseif ~cellfun('isempty',strfind(M{kk},'matrix size [1]'))
        koko = cell2mat(strfind(M{kk},':='));
        apu = cell2mat(M{kk});
        n_dim1 = str2double(apu(koko+2:end));
    elseif ~cellfun('isempty',strfind(M{kk},'matrix size [2]'))
        koko = cell2mat(strfind(M{kk},':='));
        apu = cell2mat(M{kk});
        n_dim2 = str2double(apu(koko+2:end));
    elseif ~cellfun('isempty',strfind(M{kk},'matrix size [3]'))
        koko = cell2mat(strfind(M{kk},':='));
        apu = cell2mat(M{kk});
        n_dim3 = str2double(apu(koko+2:end));
    elseif ~cellfun('isempty',strfind(M{kk},'matrix size [4]'))
        koko = cell2mat(strfind(M{kk},':='));
        apu = cell2mat(M{kk});
        n_dim4 = str2double(apu(koko+2:end));
    elseif ~cellfun('isempty',strfind(M{kk},'matrix size [5]'))
        koko = cell2mat(strfind(M{kk},':='));
        apu = cell2mat(M{kk});
        n_dim5 = str2double(apu(koko+2:end));
    elseif ~cellfun('isempty',strfind(M{kk},'imagedata byte order')) || ~cellfun('isempty',strfind(M{kk},'image data byte order'))
        koko = cell2mat(strfind(M{kk},':='));
        apu = cell2mat(M{kk});
        apu = apu(koko+2:end);
        if any(strfind(apu,'LITTLEENDIAN'))
            machinefmt = 'l';
        elseif any(strfind(apu,'BIGENDIAN'))
            machinefmt = 'b';
        end
    elseif ~cellfun('isempty',strfind(M{kk},'name of data file'))
        koko = cell2mat(strfind(M{kk},':='));
        apu = cell2mat(M{kk});
        f_name = (apu(koko+2:end));
        f_name = f_name(strfind(f_name, ' ') + 1 : end);
        if any(strfind(f_name, '/'))
            f_name = f_name(strfind(f_name, '/') + 1 : end);
        end
        [filepath] = fileparts(filename);
        if ~isempty(filepath)
            f_name = [fileparts(filename) '/' f_name];
        end
    end
end

if n_dim3 == 1
    for kk = 1 : length(M)
        if ~cellfun('isempty',strfind(M{kk},'number of slices'))
            koko = cell2mat(strfind(M{kk},':='));
            apu = cell2mat(M{kk});
            n_dim3 = str2double(apu(koko+2:end));
        end
    end
    if n_dim3 == 1 && n_dim4 == 1 && n_dim5 == 1
        for kk = 1 : length(M)
            if ~cellfun('isempty',strfind(M{kk},'total number of images'))
                koko = cell2mat(strfind(M{kk},':='));
                apu = cell2mat(M{kk});
                n_dim3 = str2double(apu(koko+2:end));
            end
        end
    end
end


if strcmp(filename(end-2:end),'hdr') || strcmp(filename(end-2:end),'h33')
    filename = [filename(1:end-3) 'img'];
end

fid = fopen(filename);
if fid == - 1
    filename = [filename(1:end-3) 'i33'];
    fid = fopen(filename);
    if fid == -1
        fid = fopen(f_name);
        if fid == -1
            error(['Could not find either ' filename ' or ' [filename(1:end-3) 'img or ' f_name]])
        end
    end
end
output = fread(fid, inf, [type '=>' type], 0, machinefmt);
try
    output = reshape(output, n_dim1, n_dim2, n_dim3, n_dim4, n_dim5);
catch
    output = reshape(output, n_dim1, n_dim2, []);
end
fclose(fid);

File: source/Python/omegatomo/io/loadInterfile.py
# -*- coding: utf-8 -*-
"""
Created on Thu Mar  7 13:36:22 2024

@author: Mad Gigerdi
"""
import re
import numpy as np
import os

def loadInterfile(filename):
    with open(filename, 'r') as f:
        lines = f.readlines()

    n_dim1 = 1
    n_dim2 = 1
    n_dim3 = 1
    n_dim4 = 1
    n_dim5 = 1
    type = 'uint16'
    machinefmt = 'l'

    for line in lines:
        if 'number format' in line:
            if 'float' in line or 'short float' in line:
                type = 'float32'
            elif 'double' in line or 'long float' in line:
                type = 'float64'
            elif 'unsigned integer' in line:
                for line2 in lines:
                    if 'number of bytes per pixel' in line2:
                        joku = re.search(r'\d+', line2).group()
                if joku == '2':
                    joku = '16'
                elif joku == '1':
                    joku = '8'
                elif joku == '4':
                    joku = '32'
                else:
                    joku = '64'
                type = 'uint' + joku
            elif 'signed integer' in line:
                for line2 in lines:
                    if 'number of bytes per pixel' in line2:
                        joku = re.search(r'\d+', line2).group()
                if joku == '2':
                    joku = '16'
                elif joku == '1':
                    joku = '8'
                elif joku == '4':
                    joku = '32'
                else:
                    joku = '64'
                type = 'int' + joku
        elif 'matrix size[1]' in line or 'matrix size [1]' in line:
            n_dim1 = int(re.findall(r'\d+', line)[1])
        elif 'matrix size[2]' in line or 'matrix size [2]' in line:
            n_dim2 = int(re.findall(r'\d+', line)[1])
        elif 'matrix size[3]' in line or 'matrix size [3]' in line:
            n_dim3 = int(re.findall(r'\d+', line)[1])
        elif 'matrix size[4]' in line or 'matrix size [4]' in line:
            n_dim4 = int(re.findall(r'\d+', line)[1])
        elif 'matrix size[5]' in line or 'matrix size [5]' in line:
            n_dim5 = int(re.findall(r'\d+', line)[1])
        elif 'imagedata byte order' in line or 'image data byte order' in line:
            if 'LITTLEENDIAN' in line:
                machinefmt = 'l'
            elif 'BIGENDIAN' in line:
                machinefmt = 'b'
        elif 'name of data file' in line:
            f_name = line.split('=')[1].strip()
            if '/' in f_name:
                f_name = f_name.split('/')[-1]

    if n_dim3 == 1:
        for line in lines:
            if 'number of slices' in line:
                n_dim3 = int(re.findall(r'\d+', line)[0])
                break
        if n_dim3 == 1 and n_dim4 == 1 and n_dim5 == 1:
            for line in lines:
                if 'total number of images' in line:
                    n_dim3 = int(re.findall(r'\d+', line)[0])
                    break

    if filename.endswith('.hdr') or filename.endswith('.h33'):
        filename = filename[:-3] + 'img'
    path = os.path.split(filename)[0] + '/'

    try:
        with open(path + f_name, 'rb') as f:
            output = np.fromfile(f, dtype=type, count=-1)
    except FileNotFoundError: 
        with open(filename, 'rb') as f:
            output = np.fromfile(f, dtype=type, count=-1)
    
    try:
        output = np.reshape(output, (n_dim1, n_dim2, n_dim3, n_dim4, n_dim5), order='F')
    except ValueError:
        output = np.reshape(output, (n_dim1, n_dim2, -1), order='F')
    
    return output